### ISSUE TITLE
katib - failed to get studyConf

### ISSUE THREAD

**Issue by DurivetMatthias** [kubeflow/katib#437]
# mnist on katib

I am trying to run a very simple hyperparameter search on katib. I use a basic mnist model.
I have deployed kubeflow using the https://deploy.kubeflow.cloud/#/deploy click-to-deploy.
Its deployed on a google kubernetes engine in a cluster called full-kfmatthias
katib recognizes the created studyjob in the ui but never shows the results.
In my case i'm trying to see how params affect the loss.
![training_output](https://user-images.githubusercontent.com/26407191/54927098-d0585900-4f11-11e9-9896-11fe008da07c.png)
the
batch1&batch2 is something I found in <https://github.com/kubeflow/katib/blob/master/cmd/metricscollector/main.go>
the above is the output of train.py, I build and push a dockerimage to my registry.

My studyjob jaml file looks like this:

    apiVersion: kubeflow.org/v1alpha1
    kind: StudyJob
    metadata:
      namespace: kubeflow
      labels:
        controller-tools.k8s.io: "1.0"
      name: mnist-job
    spec:
      studyName: mnist
      owner: matthias
      optimizationtype: minimize
      objectivevaluename: "loss"
      optimizationgoal: 0.1
      metricsnames:
        - loss
      parameterconfigs:
        - name: --loss_fn
          parametertype: categorical
          feasible:
            list:
              - sparse_categorical_crossentropy
              - mse
        - name: --optimizer
          parametertype: categorical
          feasible:
            list:
              - adam
              - sgd
              - rmsprop
      requestcount: 4
      suggestionSpec:
        suggestionAlgorithm: "random"
        requestNumber: 3
      workerspec:
        goTemplate:
          rawTemplate: |-
            apiVersion: batch/v1
            kind: Job
            metadata:
              name: {{.WorkerID}}
              namespace: kubeflow
            spec:
              template:
                spec:
                  containers:
                  - name: {{.WorkerID}}
                    image: <the correct path to my docker image>
                    command:
                    - "python"
                    - "/train.py"
                    {{- with .HyperParameters}}
                    {{- range .}}
                     - "{{.Name}}={{.Value}}"
                     {{- end}}
                     {{- end}}
                   restartPolicy: Never
      metricsCollectorSpec:
        goTemplate:
          rawTemplate: |-
        apiVersion: batch/v1beta1
        kind: CronJob
        metadata:
          name: {{.WorkerID}}
          namespace: kubeflow
        spec:
          schedule: "*/1 * * * *"
          successfulJobsHistoryLimit: 0
          failedJobsHistoryLimit: 1
          jobTemplate:
            spec:
              template:
                spec:
                  serviceAccountName: metrics-collector
                  containers:
                  - name: {{.WorkerID}}
                    image: katib/metrics-collector
                    args:
                    - "./metricscollector"
                    - "-s"
                    - "{{.StudyID}}"
                    - "-t"
                    - "{{.TrialID}}"
                    - "-w"
                    - "{{.WorkerID}}"
                    - "-n"
                    - "{{.NameSpace}}"
                  restartPolicy: Never

I noticed that most tutorials/github issues only include this yaml and never actually show any python code or dockerfile to build the images. How does the metrics-collector get the metrics from the stdout of my docker image? It also seems like the metrics-collector is never being created.

I am able to run the random-example.
I also noticed the random-example does not have a definition for a metrics-collector. does this mean there's a way to directly return the metrics? or is there a default metrics-collector?

## debug prints

![afbeelding](https://user-images.githubusercontent.com/26407191/54929780-67271480-4f16-11e9-884e-5c746f19f35a.png)

![afbeelding](https://user-images.githubusercontent.com/26407191/54930185-1f54bd00-4f17-11e9-9dbd-53b58a62f7c5.png)

then it creates and fails 3 more times and finishes with
`Study reached request count. It is completed.`

![afbeelding](https://user-images.githubusercontent.com/26407191/54930263-47442080-4f17-11e9-8e5e-2a246d1b9e08.png)

`kubectl describe studyjob mnist-job -n kubeflow` results in:

```bash
Name:         mnist-job
Namespace:    kubeflow
Labels:       controller-tools.k8s.io=1.0
Annotations:  <none>
API Version:  kubeflow.org/v1alpha1
Kind:         StudyJob
Metadata:
  Creation Timestamp:  2019-03-25T14:52:34Z
  Generation:          1
  Resource Version:    7565
  Self Link:           /apis/kubeflow.org/v1alpha1/namespaces/kubeflow/studyjobs/mnist-job
  UID:                 9fb74f1c-4f0d-11e9-862a-42010a84002f
Spec:
  Metrics Collector Spec:
    Go Template:
      Raw Template:  apiVersion: batch/v1beta1
kind: CronJob
metadata:
  name: {{.WorkerID}}
  namespace: kubeflow
spec:
  schedule: "*/1 * * * *"
  successfulJobsHistoryLimit: 1
  failedJobsHistoryLimit: 1
  jobTemplate:
    spec:
      template:
        spec:
          serviceAccountName: metrics-collector
          containers:
          - name: {{.WorkerID}}
            image: katib/metrics-collector
            args:
            - "./metricscollector"
            - "-s"
            - "{{.StudyID}}"
            - "-t"
            - "{{.TrialID}}"
            - "-w"
            - "{{.WorkerID}}"
            - "-n"
            - "{{.NameSpace}}"
          restartPolicy: Never
    Metricsnames:
        loss
    Objectivevaluename:  loss
    Optimizationgoal:    0
    Optimizationtype:    minimize
    Owner:               matthias
    Parameterconfigs:
        Feasible:
        List:
            adam
            sgd
            rmsprop
        Name:           --optimizer
        Parametertype:  categorical
    Requestcount:     4
    Study Name:       mnist
    Suggestion Spec:
        Request Number:        3
        Suggestion Algorithm:  random
        Suggestion Parameters:
        Name:   SuggestionCount
        Value:  0
    Worker Spec:
        Go Template:
        Raw Template:  apiVersion: batch/v1
kind: Job
metadata:
  name: {{.WorkerID}}
  namespace: kubeflow
spec:
  template:
    spec:
      containers:
      - name: {{.WorkerID}}
        image: eu.gcr.io/ml6-sandbox/kfmatthias/katib/mnist
        command:
        - "python"
        - "/train.py"
        {{- with .HyperParameters}}
        {{- range .}}
        - "{{.Name}}={{.Value}}"
        {{- end}}
        {{- end}}
      restartPolicy: Never
Status:
  Completion Time:              2019-03-25T14:53:28Z
  Conditon:                     Completed
  Early Stopping Parameter Id:
  Last Reconcile Time:          2019-03-25T14:53:28Z
  Start Time:                   2019-03-25T14:52:34Z
  Studyid:                      y5446df3d81260a8
  Suggestion Count:             4
  Suggestion Parameter Id:      ne47d45e086b05fb
  Trials:
    Trialid:  o28c37317869fb60
    Workeridlist:
      Completion Time:  <nil>
      Conditon:         Failed
      Kind:             Job
      Start Time:       2019-03-25T14:52:34Z
      Workerid:         xa4ddc2496aeef71
    Trialid:            gfa8386222343837
    Workeridlist:
      Completion Time:  <nil>
      Conditon:         Failed
      Kind:             Job
      Start Time:       2019-03-25T14:52:34Z
      Workerid:         l71e9b510aae56c0
    Trialid:            hfa2b891087e28d5
    Workeridlist:
      Completion Time:  <nil>
      Conditon:         Failed
      Kind:             Job
      Start Time:       2019-03-25T14:52:34Z
      Workerid:         l8b8c42ea251c5c7
    Trialid:            ga0f239c9d16eb87
    Workeridlist:
      Completion Time:  <nil>
      Conditon:         Failed
      Kind:             Job
      Start Time:       2019-03-25T14:53:09Z
      Workerid:         k7cc4f2be6fd891b
    Trialid:            v079de176a3518b0
    Workeridlist:
      Completion Time:  <nil>
      Conditon:         Failed
      Kind:             Job
      Start Time:       2019-03-25T14:53:09Z
      Workerid:         s55a9c141c7a7dae
    Trialid:            xb427617d1cd5519
    Workeridlist:
      Completion Time:  <nil>
      Conditon:         Failed
      Kind:             Job
      Start Time:       2019-03-25T14:53:09Z
      Workerid:         ld5dcc08de6fc44b
    Trialid:            v177289cdeb76fee
    Workeridlist:
      Completion Time:  <nil>
      Conditon:         Failed
      Kind:             Job
      Start Time:       2019-03-25T14:53:17Z
      Workerid:         u6662ceb72133a4c
    Trialid:            f1d06fb42e8c7b0a
    Workeridlist:
      Completion Time:  <nil>
      Conditon:         Failed
      Kind:             Job
      Start Time:       2019-03-25T14:53:17Z
      Workerid:         z3cffc44194b3358
    Trialid:            v33c23e8303f3ce7
    Workeridlist:
      Completion Time:  <nil>
      Conditon:         Failed
      Kind:             Job
      Start Time:       2019-03-25T14:53:17Z
      Workerid:         z2ca5624c3856f46
    Trialid:            s193effdeaa7ebda
    Workeridlist:
      Completion Time:  <nil>
      Conditon:         Failed
      Kind:             Job
      Start Time:       2019-03-25T14:53:22Z
      Workerid:         z62e5902ef218aca
    Trialid:            x122390876ae701e
    Workeridlist:
      Completion Time:  <nil>
      Conditon:         Failed
      Kind:             Job
      Start Time:       2019-03-25T14:53:22Z
      Workerid:         a268b2a0ce8f6f38
    Trialid:            k385b978cadb6259
    Workeridlist:
      Completion Time:  <nil>
      Conditon:         Failed
      Kind:             Job
      Start Time:       2019-03-25T14:53:22Z
      Workerid:         k0e3392d0b6f5243
Events:                 <none>
```

**Comment by YujiOshima** [comment]
Please add the `-m` option to metrics collector template.
like below.

```
  metricsCollectorSpec:
    goTemplate:
      rawTemplate: |-
    apiVersion: batch/v1beta1
    kind: CronJob
    metadata:
      name: {{.WorkerID}}
      namespace: kubeflow
    spec:
      schedule: "*/1 * * * *"
      successfulJobsHistoryLimit: 0
      failedJobsHistoryLimit: 1
      jobTemplate:
        spec:
          template:
            spec:
              serviceAccountName: metrics-collector
              containers:
              - name: {{.WorkerID}}
                image: katib/metrics-collector
                args:
                - "./metricscollector"
                - "-s"
                - "{{.StudyID}}"
                - "-t"
                - "{{.TrialID}}"
                - "-w"
                - "{{.WorkerID}}"
                - "-n"
                - "{{.NameSpace}}"
                - "-m"
                - "{{.ManagerSerivce}}"
              restartPolicy: Never
```

**Comment by DurivetMatthias** [comment]
I added:
```yaml
- "-m"
- "{{.ManagerSerivce}}"
```
I can't see any failed pods in my kubernetes cluster but these errors appear in the studyjob-controller:

![afbeelding](https://user-images.githubusercontent.com/26407191/55063245-dd8f5800-5077-11e9-93dd-831e19e6ad68.png)

![afbeelding](https://user-images.githubusercontent.com/26407191/55063210-c94b5b00-5077-11e9-92d5-7087ad076bfe.png)

in the last image the pod with job-name=n101a90ea6849a42 is deleted
and right after that, the studyjob complains that it can't find the pod?
(this happens multiple times with different pods in the logs above it)

Can you explain exactly how the metrics collecting works? or is there some documentation on it somewhere? @YujiOshima 

(I suppose the -m option did help somewhat)
